src/lib/structures/Josh.ts

```diff
@@ -0,0 +1,1813 @@
+    if (optionsOrInstance instanceof Middleware) this.middlewares.set(optionsOrInstance.name, optionsOrInstance);
+    else {
+      if (hook === undefined)
+        throw new JoshError({
+          identifier: Josh.Identifiers.UseMiddlewareHookNotFound,
+          message: 'The "hook" parameter for middleware was not found.'
+        });
+
+      const { name, position, trigger, method } = optionsOrInstance;
+      const options: Middleware.Options = { name, position, conditions: { pre: [], post: [] } };
+      const middleware = this.middlewares.get(options.name) ?? new Middleware<StoredValue>(options).init(this.middlewares);
+
+      if (trigger !== undefined && method !== undefined) options.conditions[trigger === Trigger.PreProvider ? 'pre' : 'post'].push(method);
+
+      Object.defineProperty(middleware, method === undefined ? 'run' : method, { value: hook });
+
+      this.middlewares.set(middleware.name, middleware);
+    }
+
+    return this;
+  }
+
+  /**
+   * Generate an automatic key. Generally an integer incremented by `1`, but depends on provider.
+   * @since 2.0.0
+   * @returns The newly generated automatic key.
+   *
+   * @example
+   * ```typescript
+   * const key = await josh.autoKey();
+   *
+   * await josh.set(key, 'value');
+   * ```
+   */
+  public async autoKey(): Promise<string> {
+    let payload: AutoKeyPayload = { method: Method.AutoKey, trigger: Trigger.PreProvider, data: '' };
+
+    for (const middleware of this.middlewares.array()) await middleware.run(payload);
+    for (const middleware of this.middlewares.getPreMiddlewares(Method.AutoKey)) payload = await middleware[Method.AutoKey](payload);
+
+    payload = await this.provider[Method.AutoKey](payload);
+    payload.trigger = Trigger.PostProvider;
+
+    if (payload.error) throw payload.error;
+
+    for (const middleware of this.middlewares.array()) await middleware.run(payload);
+    for (const middleware of this.middlewares.getPostMiddlewares(Method.AutoKey)) payload = await middleware[Method.AutoKey](payload);
+
+    return payload.data;
+  }
+
+  /**
+   * Clears all stored values from the provider.
+   * @since 2.0.0
+   * @returns The {@link Josh} instance.
+   *
+   * @example
+   * ```typescript
+   * await josh.set('key', 'value');
+   *
+   * await josh.clear();
+   *
+   * await josh.get('key'); // null
+   * ```
+   */
+  public async clear(): Promise<this> {
+    let payload: ClearPayload = { method: Method.Clear, trigger: Trigger.PreProvider };
+
+    for (const middleware of this.middlewares.array()) await middleware.run(payload);
+    for (const middleware of this.middlewares.getPreMiddlewares(Method.Clear)) payload = await middleware[Method.Clear](payload);
+
+    payload = await this.provider[Method.Clear](payload);
+    payload.trigger = Trigger.PostProvider;
+
+    if (payload.error) throw payload.error;
+
+    for (const middleware of this.middlewares.array()) await middleware.run(payload);
```

Comment:
wait why are all external middleware being run before and after? Is there no concept of pre/post middleware that are user-provided? I wouldn't expect this to run all my middleware twice, unless I'm misunderstanding something.

Reply:
Yes there is pre and post middleware. Payloads have a `trigger` property which tell you which state it's in, plus when you add a middleware you tell Josh when to run it.